File: utils/todoParser.ts
export interface ParsedDates {
  creationDate: string;
  completionDate?: string;
  dueDate?: string;
  time?: string;
}

export function parseDatesFromRaw(raw: string, fallbackCreation?: string): ParsedDates {
  let creationDate = fallbackCreation || new Date().toISOString().split('T')[0];
  let completionDate: string | undefined = undefined;
  let dueDate: string | undefined = undefined;
  let time: string | undefined = undefined;

  // Time tag: time:HH:MM or @HH:MM (e.g. time:14:30)
  const timeMatch = raw.match(/\btime:(\d{1,2}:\d{2})\b/i);
  if (timeMatch) {
    time = timeMatch[1].padStart(5, '0');
  }

  // Due date: due:YYYY-MM-DD or due:YYYY-MM-DDTHH:MM
  const dueMatch = raw.match(/\bdue:(\d{4}-\d{2}-\d{2})(?:T(\d{1,2}:\d{2}))?\b/i);
  if (dueMatch) {
    dueDate = dueMatch[1];
    if (dueMatch[2]) {
      time = dueMatch[2].padStart(5, '0');
    }
  }

  // Check if completed: x YYYY-MM-DD YYYY-MM-DD ...
  const completedWithTwoDatesMatch = raw.match(/^x\s+(\d{4}-\d{2}-\d{2})\s+(\d{4}-\d{2}-\d{2})\s/);
  if (completedWithTwoDatesMatch) {
    completionDate = completedWithTwoDatesMatch[1];
    creationDate = completedWithTwoDatesMatch[2];
    return { creationDate, completionDate, dueDate, time };
  }

  // Priority date: (A) YYYY-MM-DD ...
  const priorityDateMatch = raw.match(/^\([A-Z]\)\s+(\d{4}-\d{2}-\d{2})\s/);
  if (priorityDateMatch) {
    creationDate = priorityDateMatch[1];
    return { creationDate, completionDate, dueDate, time };
  }

  // Plain creation date at start: YYYY-MM-DD ...
  const plainDateMatch = raw.match(/^(\d{4}-\d{2}-\d{2})\s/);
  if (plainDateMatch) {
    creationDate = plainDateMatch[1];
    return { creationDate, completionDate, dueDate, time };
  }

  return { creationDate, completionDate, dueDate, time };
}

export function updateRawDates(
  raw: string,
  newCreationDate?: string,
  newDueDate?: string | null,
  newTime?: string | null
): string {
  let updatedRaw = raw;

  // 1. Update or clear time (time:HH:MM)
  if (newTime !== undefined) {
    if (newTime && newTime.trim() !== '') {
      if (/\btime:\d{1,2}:\d{2}\b/i.test(updatedRaw)) {
        updatedRaw = updatedRaw.replace(/\btime:\d{1,2}:\d{2}\b/i, `time:${newTime}`);
      } else {
        updatedRaw = `${updatedRaw.trim()} time:${newTime}`;
      }
    } else {
      updatedRaw = updatedRaw.replace(/\s*\btime:\d{1,2}:\d{2}\b/i, '').trim();
    }
  }

  // 2. Update or clear due date (due:YYYY-MM-DD)
  if (newDueDate !== undefined) {
    if (newDueDate && newDueDate.trim() !== '') {
      if (/\bdue:\d{4}-\d{2}-\d{2}(?:T\d{1,2}:\d{2})?\b/i.test(updatedRaw)) {
        updatedRaw = updatedRaw.replace(/\bdue:\d{4}-\d{2}-\d{2}(?:T\d{1,2}:\d{2})?\b/i, `due:${newDueDate}`);
      } else {
        updatedRaw = `${updatedRaw.trim()} due:${newDueDate}`;
      }
    } else {
      // Clear due date
      updatedRaw = updatedRaw.replace(/\s*\bdue:\d{4}-\d{2}-\d{2}(?:T\d{1,2}:\d{2})?\b/i, '').trim();
    }
  }

  // 3. Update creation date if provided
  if (newCreationDate && /^\d{4}-\d{2}-\d{2}$/.test(newCreationDate)) {
    if (/^x\s+(\d{4}-\d{2}-\d{2})\s+(\d{4}-\d{2}-\d{2})\s/.test(updatedRaw)) {
      updatedRaw = updatedRaw.replace(/^x\s+(\d{4}-\d{2}-\d{2})\s+(\d{4}-\d{2}-\d{2})\s/, `x $1 ${newCreationDate} `);
    } else if (/^\([A-Z]\)\s+(\d{4}-\d{2}-\d{2})\s/.test(updatedRaw)) {
      updatedRaw = updatedRaw.replace(/^(\([A-Z]\))\s+(\d{4}-\d{2}-\d{2})\s/, `$1 ${newCreationDate} `);
    } else if (/^(\d{4}-\d{2}-\d{2})\s/.test(updatedRaw)) {
      updatedRaw = updatedRaw.replace(/^(\d{4}-\d{2}-\d{2})\s/, `${newCreationDate} `);
    } else if (/^\([A-Z]\)\s/.test(updatedRaw)) {
      updatedRaw = updatedRaw.replace(/^(\([A-Z]\))\s/, `$1 ${newCreationDate} `);
    } else if (!updatedRaw.startsWith('x ')) {
      updatedRaw = `${newCreationDate} ${updatedRaw}`;
    }
  }

  return updatedRaw;
}

File: flutter_app/lib/utils/todo_parser.dart
import 'package:intl/intl.dart';

class ParsedTaskMeta {
  final String title;
  final String? priority;
  final String creationDate;
  final String? completionDate;
  final String? dueDate;
  final String? dueTime;
  final String? recurrence;
  final bool completed;
  final List<String> projects;
  final List<String> contexts;

  ParsedTaskMeta({
    required this.title,
    this.priority,
    required this.creationDate,
    this.completionDate,
    this.dueDate,
    this.dueTime,
    this.recurrence,
    required this.completed,
    required this.projects,
    required this.contexts,
  });
}

ParsedTaskMeta parseRawToStructured(String raw, [String? fallbackCreation]) {
  final words = raw.trim().split(RegExp(r'\s+'));
  if (words.isEmpty || words.first.isEmpty) {
    final today = fallbackCreation ?? DateFormat('yyyy-MM-dd').format(DateTime.now());
    return ParsedTaskMeta(
      title: '',
      creationDate: today,
      completed: false,
      projects: [],
      contexts: [],
    );
  }

  bool completed = false;
  String? priority;
  String creationDate = fallbackCreation ?? DateFormat('yyyy-MM-dd').format(DateTime.now());
  String? completionDate;
  String? dueDate;
  String? dueTime;
  String? recurrence;
  final List<String> projects = [];
  final List<String> contexts = [];
  final List<String> titleWords = [];

  int idx = 0;
  final dateRegex = RegExp(r'^\d{4}-\d{2}-\d{2}$');

  // Check completed prefix: x YYYY-MM-DD YYYY-MM-DD ... or x ...
  if (words[0] == 'x') {
    completed = true;
    idx++;
    if (idx < words.length && dateRegex.hasMatch(words[idx])) {
      completionDate = words[idx];
      idx++;
      if (idx < words.length && dateRegex.hasMatch(words[idx])) {
        creationDate = words[idx];
        idx++;
      }
    }
  }

  // Check priority: (A)
  if (idx < words.length && RegExp(r'^\([A-Z]\)$').hasMatch(words[idx])) {
    priority = words[idx][1];
    idx++;
  }

  // Check creation date if not completed: YYYY-MM-DD
  if (!completed && idx < words.length && dateRegex.hasMatch(words[idx])) {
    creationDate = words[idx];
    idx++;
  }

  // Process remaining tokens
  for (; idx < words.length; idx++) {
    final word = words[idx];

    if (word.startsWith('+') && word.length > 1) {
      if (!projects.contains(word)) projects.add(word);
    } else if (word.startsWith('@') && word.length > 1) {
      if (!contexts.contains(word)) contexts.add(word);
    } else if (RegExp(r'\bdue:(\d{4}-\d{2}-\d{2})(?:T(\d{1,2}:\d{2}))?\b', caseSensitive: false).hasMatch(word)) {
      final match = RegExp(r'\bdue:(\d{4}-\d{2}-\d{2})(?:T(\d{1,2}:\d{2}))?\b', caseSensitive: false).firstMatch(word);
      if (match != null) {
        dueDate = match.group(1);
        if (match.group(2) != null) dueTime = match.group(2)!.padLeft(5, '0');
      }
    } else if (RegExp(r'\btime:(\d{1,2}:\d{2})\b', caseSensitive: false).hasMatch(word)) {
      final match = RegExp(r'\btime:(\d{1,2}:\d{2})\b', caseSensitive: false).firstMatch(word);
      if (match != null) dueTime = match.group(1)!.padLeft(5, '0');
    } else if (RegExp(r'\brec:(?:strict:|\+)?(?:\d+[dwmy]|weekday|mwf)\b', caseSensitive: false).hasMatch(word)) {
      final match = RegExp(r'\brec:((?:strict:|\+)?(?:\d+[dwmy]|weekday|mwf))\b', caseSensitive: false).firstMatch(word);
      if (match != null) recurrence = match.group(1);
    } else {
      titleWords.add(word);
    }
  }

  final title = titleWords.isNotEmpty ? titleWords.join(' ') : raw;

  return ParsedTaskMeta(
    title: title,
    priority: priority,
    creationDate: creationDate,
    completionDate: completionDate,
    dueDate: dueDate,
    dueTime: dueTime,
    recurrence: recurrence,
    completed: completed,
    projects: projects,
    contexts: contexts,
  );
}

String buildRawFromStructured({
  required String title,
  String? priority,
  required String creationDate,
  String? completionDate,
  String? dueDate,
  String? dueTime,
  String? recurrence,
  required bool completed,
  List<String>? projects,
  List<String>? contexts,
}) {
  final List<String> parts = [];

  if (completed) {
    parts.add('x');
    if (completionDate != null && completionDate.isNotEmpty) parts.add(completionDate);
    parts.add(creationDate);
  }

  if (priority != null && priority.isNotEmpty) {
    parts.add('($priority)');
  }

  if (!completed && creationDate.isNotEmpty) {
    parts.add(creationDate);
  }

  if (title.isNotEmpty) {
    parts.add(title);
  }

  if (projects != null) {
    for (final p in projects) {
      if (!parts.contains(p)) parts.add(p);
    }
  }

  if (contexts != null) {
    for (final c in contexts) {
      if (!parts.contains(c)) parts.add(c);
    }
  }

  if (dueDate != null && dueDate.isNotEmpty) {
    parts.add('due:$dueDate');
  }

  if (dueTime != null && dueTime.isNotEmpty) {
    parts.add('time:$dueTime');
  }

  if (recurrence != null && recurrence.isNotEmpty) {
    final recStr = recurrence.startsWith('rec:') ? recurrence : 'rec:$recurrence';
    parts.add(recStr);
  }

  return parts.join(' ');
}

String updateRawDates(
  String raw, {
  String? newCreationDate,
  String? newDueDate,
  bool clearDueDate = false,
  String? newTime,
  bool clearTime = false,
}) {
  final parsed = parseRawToStructured(raw, newCreationDate);
  return buildRawFromStructured(
    title: parsed.title,
    priority: parsed.priority,
    completed: parsed.completed,
    completionDate: parsed.completionDate,
    creationDate: newCreationDate ?? parsed.creationDate,
    dueDate: clearDueDate ? null : (newDueDate ?? parsed.dueDate),
    dueTime: clearTime ? null : (newTime ?? parsed.dueTime),
    recurrence: parsed.recurrence,
    projects: parsed.projects,
    contexts: parsed.contexts,
  );
}
